docs/tutorials/go-built-in.md

```diff
@@ -56,338 +56,495 @@ We'll start by creating a new Go project.
+go get github.com/cometbft/cometbft@v0.34.24
+```
 
-When a new transaction is added to the Tendermint Core, it will ask the
-application to check it (validate the format, signatures, etc.).
+Now go back to the `main.go` and modify the `main` function so it matches the following,
+where an instance of the `KVStoreApplication` type is created.
 
 ```go
-import "bytes"
+func main() {
+    fmt.Println("Hello, CometBFT")
+
+    _ = NewKVStoreApplication()
+}
+```
+
+You can recompile and run the application now by running `go get` and `go build`, but it does
+not do anything.
+So let's revisit the code adding the logic needed to implement our minimal key/value store
+and to start it along with the CometBFT Service.
+
+
+### 1.3.1 Add a persistent data store
 
-func (app *KVStoreApplication) isValid(tx []byte) (code uint32) {
+Our application will need to write its state out to persistent storage so that it
+can stop and start without losing all of its data.
+
+For this tutorial, we will use [BadgerDB](https://github.com/dgraph-io/badger), a
+a fast embedded key-value store.
+
+First, add Badger as a dependency of your go module using the `go get` command:
+
+`go get github.com/dgraph-io/badger/v3`
+
+Next, let's update the application and its constructor to receive a handle to the database, as follows:
+
+```go
+type KVStoreApplication struct {
+	db           *badger.DB
+	onGoingBlock *badger.Txn
+}
+
+var _ abcitypes.Application = (*KVStoreApplication)(nil)
+
+func NewKVStoreApplication(db *badger.DB) *KVStoreApplication {
+	return &KVStoreApplication{db: db}
+}
+```
+
+The `onGoingBlock` keeps track of the Badger transaction that will update the application's state when a block
+is completed. Don't worry about it for now, we'll get to that later.
+
+Next, update the `import` stanza at the top to include the Badger library:
+
+```go
+import(
+	"github.com/dgraph-io/badger/v3"
+	abcitypes "github.com/cometbft/cometbft/abci/types"
+)
+```
+
+Finally, update the `main.go` file to invoke the updated constructor:
+
+```go
+	_ = NewKVStoreApplication(nil)
+```
+
+### 1.3.2 CheckTx
+
+When CometBFT receives a new transaction from a client, or from another full node,
+CometBFT asks the application if the transaction is acceptable, using the `CheckTx` method.
+Invalid transactions will not be shared with other nodes and will not become part of any blocks and, therefore, will not be executed by the application.
+
+In our application, a transaction is a string with the form `key=value`, indicating a key and value to write to the store.
+
+The most basic validation check we can perform is to check if the transaction conforms to the `key=value` pattern.
+For that, let's add the following helper method to app.go:
+
+```go
+func (app *KVStoreApplication) isValid(tx []byte) uint32 {
 	// check format
 	parts := bytes.Split(tx, []byte("="))
 	if len(parts) != 2 {
 		return 1
 	}
 
-	key, value := parts[0], parts[1]
-
-	// check if the same key=value already exists
```

Comment:
Is this removed in the code too?

Reply:
The new version of the tutorial doesn't include this more comprehensive check...

But which code are you referring to?

Reply:
Sorry, I thought this was actually written as some example somewhere in the code as well. My bad.